File: modules/uc-engine/ios/NativeRuntimeDiagnostics.swift
import Foundation
import Darwin

@_silgen_name("flock")
private func diagnosticFlock(_ descriptor: Int32, _ operation: Int32) -> Int32

public enum NativeDiagnosticRole: String, Sendable {
  case mainApplication = "main"
  case keyboardExtension = "keyboard"
  case shareExtension = "share"
}

public enum NativeDiagnosticEvent: String, Sendable {
  case captureStarted = "capture.started"
  case extensionVisible = "extension.visible"
  case extensionHidden = "extension.hidden"
  case appForeground = "app.foreground"
  case appBackground = "app.background"
  case engineStart = "engine.start"
  case engineSuspend = "engine.suspend"
  case engineResume = "engine.resume"
  case engineShutdown = "engine.shutdown"
  case secureStateRecovery = "security.recover"
  case ownershipAcquire = "ownership.acquire"
  case ownershipRelease = "ownership.release"
  case networkChanged = "network.path_observed"
  case networkObservation = "network.observation"
  case shareHandoff = "share.handoff"
  case exportFlush = "export.flush"
}

public enum NativeDiagnosticTrigger: String, Sendable {
  case appStartup, appForeground, appBackground, userRequest
  case extensionVisible, extensionHidden, runtimeHandoff, contextDestroyed
  case networkChange, exportRequested, unspecified
}

public enum NativeDiagnosticOutcome: String, Sendable {
  case observed, started, succeeded, failed, notAcquired
}

public struct NativeDiagnosticFailure: Encodable, Sendable {
  public enum Reason: String, Encodable, Sendable {
    case engineFailure, permissionDenied, storageUnavailable, runtimeBusy, spaceUnavailable
    case cancelled, nativeFailure
  }
  public let reason: Reason
  public let code: Int64?
  public let retryable: Bool?

  public init(reason: Reason, code: Int64? = nil, retryable: Bool? = nil) {
    self.reason = reason
    self.code = code
    self.retryable = retryable
  }
}

public struct NativeDiagnosticNetwork: Encodable, Equatable, Sendable {
  public enum Kind: String, Encodable, Sendable { case wifi, cellular, wired, other, none }
  public enum Change: String, Encodable, Sendable { case initial, pathUpdate, defaultNetwork, properties, capabilities }
  public let change: Change
  public let available: Bool
  public let kind: Kind
  public let expensive: Bool
  public let constrained: Bool

  public init(available: Bool, kind: Kind, expensive: Bool = false, constrained: Bool = false, change: Change = .capabilities) {
    self.change = change
    self.available = available
    self.kind = kind
    self.expensive = expensive
    self.constrained = constrained
  }
}

public struct NativeDiagnosticOperation: Sendable {
  fileprivate let id: String
  fileprivate let event: NativeDiagnosticEvent
  fileprivate let trigger: NativeDiagnosticTrigger
  fileprivate let startedAt: UInt64
  fileprivate let generation: UInt64
}

public struct NativeDiagnosticsSnapshot: Sendable {
  public let writerStatus: String
  public let droppedRecords: UInt64
  public let writeFailures: UInt64
  public let prunedFiles: UInt64
  public let sessionId: String
  public let role: String
  public let startedAt: String
  public let capturedAt: String
  public let maxFileBytes: Int
  public let maxFiles: Int

  public var dictionary: [String: Any] {
    [
      "writerStatus": writerStatus, "droppedRecords": NSNumber(value: droppedRecords),
      "writeFailures": NSNumber(value: writeFailures), "prunedFiles": NSNumber(value: prunedFiles),
      "sessionId": sessionId, "role": role, "startedAt": startedAt, "capturedAt": capturedAt,
      "policy": "native-runtime-boundaries-v1", "effectiveLevel": "info",
      "filteredRecordCount": NSNull(), "retentionDays": 3,
      "maxFileBytes": maxFileBytes, "maxFiles": maxFiles,
    ]
  }
}

/// A closed event vocabulary deliberately excludes arbitrary messages and attributes.
/// Each process writes its own files; it never shares an open writer with an extension.
public final class NativeRuntimeDiagnostics: @unchecked Sendable {
  private let directory: URL?
  private let role: NativeDiagnosticRole
  private let sessionId = UUID().uuidString.lowercased()
  private let startedAt = Date()
  private let queue = DispatchQueue(label: "app.uniclipboard.native-diagnostics", qos: .utility)
  private let capacity = DispatchSemaphore(value: 128)
  private let state = NSLock()
  private let maxFileBytes: Int
  private let maxFiles: Int
  private var generation: UInt64 = 0
  private var droppedRecords: UInt64 = 0
  private var writeFailures: UInt64 = 0
  private var prunedFiles: UInt64 = 0
  private var writerReady = false
  private var fileIndex = 0
  private var fileBytes = 0

  public init(directory: URL?, role: NativeDiagnosticRole, maxFileBytes: Int = 256 * 1024, maxFiles: Int = 24) {
    self.directory = directory
    self.role = role
    self.maxFileBytes = max(1_024, maxFileBytes)
    self.maxFiles = max(1, maxFiles)
    record(.captureStarted, trigger: .unspecified)
  }

  public func begin(_ event: NativeDiagnosticEvent, trigger: NativeDiagnosticTrigger) -> NativeDiagnosticOperation {
    let current = state.withLock {
      if event == .engineStart { generation += 1 }
      return generation
    }
    let operation = NativeDiagnosticOperation(
      id: UUID().uuidString.lowercased(), event: event, trigger: trigger,
      startedAt: DispatchTime.now().uptimeNanoseconds, generation: current
    )
    enqueue(event, trigger: trigger, outcome: .started, operation: operation)
    return operation
  }

  public func finish(_ operation: NativeDiagnosticOperation, outcome: NativeDiagnosticOutcome,
                     failure: NativeDiagnosticFailure? = nil) {
    enqueue(operation.event, trigger: operation.trigger, outcome: outcome, operation: operation, failure: failure)
  }

  public func record(_ event: NativeDiagnosticEvent, trigger: NativeDiagnosticTrigger,
                     outcome: NativeDiagnosticOutcome = .observed,
                     failure: NativeDiagnosticFailure? = nil, network: NativeDiagnosticNetwork? = nil) {
    enqueue(event, trigger: trigger, outcome: outcome, failure: failure, network: network)
  }

  public func flush(deadlineMs: Int = 1_000) -> Bool {
    let completed = DispatchSemaphore(value: 0)
    queue.async { completed.signal() }
    return completed.wait(timeout: .now() + .milliseconds(max(0, deadlineMs))) == .success
  }

  public func snapshot() -> NativeDiagnosticsSnapshot {
    state.withLock {
      NativeDiagnosticsSnapshot(
        writerStatus: writerReady ? "ready" : "unavailable", droppedRecords: droppedRecords,
        writeFailures: writeFailures, prunedFiles: prunedFiles, sessionId: sessionId,
        role: role.rawValue, startedAt: Self.timestamp(startedAt), capturedAt: Self.timestamp(Date()),
        maxFileBytes: maxFileBytes, maxFiles: maxFiles
      )
    }
  }

  public func fileURLs() -> [URL] { fileDiscovery().files }

  public func fileDiscovery() -> (files: [URL], status: String, skippedFiles: Int) {
    guard let directory else { return ([], "unavailable", 0) }
    do {
      let candidates = try FileManager.default.contentsOfDirectory(
        at: directory, includingPropertiesForKeys: [.isRegularFileKey, .isSymbolicLinkKey],
        options: [.skipsHiddenFiles]
      ).filter { Self.isJournalFile($0.lastPathComponent) }
      var skipped = 0
      let files = candidates.filter { url in
        guard let values = try? url.resourceValues(forKeys: [.isRegularFileKey, .isSymbolicLinkKey]) else {
          skipped += 1
          return false
        }
        if values.isSymbolicLink == true { skipped += 1; return false }
        if values.isRegularFile != true { skipped += 1; return false }
        return true
      }.sorted { $0.lastPathComponent < $1.lastPathComponent }
      return (files, skipped == 0 ? "completed" : "partial", skipped)
    } catch { return ([], "unavailable", 0) }
  }

  private func enqueue(_ event: NativeDiagnosticEvent, trigger: NativeDiagnosticTrigger,
                       outcome: NativeDiagnosticOutcome, operation: NativeDiagnosticOperation? = nil,
                       failure: NativeDiagnosticFailure? = nil, network: NativeDiagnosticNetwork? = nil) {
    guard capacity.wait(timeout: .now()) == .success else {
      state.withLock { droppedRecords += 1 }
      return
    }
    let capturedAt = Date()
    let uptime = DispatchTime.now().uptimeNanoseconds
    let currentGeneration = operation?.generation ?? state.withLock { generation }
    queue.async { [self] in
      defer { capacity.signal() }
      do {
        guard let directory else { throw CocoaError(.fileNoSuchFile) }
        let health = snapshot()
        var record: [String: Any] = [
          "schemaVersion": 1, "policy": "native-runtime-boundaries-v1",
          "timestamp": Self.timestamp(capturedAt), "monotonicMs": uptime / 1_000_000,
          "role": role.rawValue, "sessionId": sessionId, "startAttempt": currentGeneration,
          "processId": ProcessInfo.processInfo.processIdentifier,
          "event": event.rawValue, "trigger": trigger.rawValue, "outcome": outcome.rawValue,
          "level": outcome == .failed ? "ERROR" : "INFO",
          "appVersion": Self.buildValue("CFBundleShortVersionString"),
          "appBuild": Self.buildValue("CFBundleVersion"),
          "droppedRecords": health.droppedRecords, "writeFailures": health.writeFailures,
          "prunedFiles": health.prunedFiles,
        ]
        if let operation {
          record["operationId"] = operation.id
          if outcome != .started { record["durationMs"] = (uptime - operation.startedAt) / 1_000_000 }
        }
        if let failure { record["failure"] = try JSONSerialization.jsonObject(with: JSONEncoder().encode(failure)) }
        if let network { record["network"] = try JSONSerialization.jsonObject(with: JSONEncoder().encode(network)) }
        var bytes = try JSONSerialization.data(withJSONObject: record, options: [.sortedKeys])
        bytes.append(0x0a)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        try withDirectoryLock(directory) {
          if fileBytes > 0 && fileBytes + bytes.count > maxFileBytes { fileIndex += 1; fileBytes = 0 }
          let url = directory.appendingPathComponent("native-runtime.\(role.rawValue).\(sessionId).\(fileIndex).jsonl")
          try prune(excluding: url)
          if !FileManager.default.fileExists(atPath: url.path) {
            var attributes: [FileAttributeKey: Any] = [.posixPermissions: 0o600]
            #if os(iOS)
            attributes[.protectionKey] = FileProtectionType.completeUntilFirstUserAuthentication
            #endif
            guard FileManager.default.createFile(atPath: url.path, contents: nil, attributes: attributes) else {
              throw CocoaError(.fileWriteUnknown)
            }
          }
          let file = try FileHandle(forWritingTo: url)
          defer { try? file.close() }
          try file.seekToEnd()
          try file.write(contentsOf: bytes)
          fileBytes += bytes.count
          state.withLock { writerReady = true }
        }
      } catch {
        state.withLock { droppedRecords += 1; writeFailures += 1; writerReady = false }
      }
    }
  }

  private func prune(excluding current: URL) throws {
    let cutoff = Date().addingTimeInterval(-3 * 24 * 60 * 60)
    let discovery = fileDiscovery()
    guard discovery.status != "unavailable" else { throw CocoaError(.fileReadUnknown) }
    let files = discovery.files.map { url in
      (url, (try? url.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast)
    }.sorted { $0.1 < $1.1 }
    var remaining = files.count + discovery.skippedFiles + (files.contains { $0.0 == current } ? 0 : 1)
    for (url, modifiedAt) in files where url != current {
      if modifiedAt < cutoff || remaining > maxFiles {
        try FileManager.default.removeItem(at: url)
        remaining -= 1
        state.withLock { prunedFiles += 1 }
      }
    }
    guard remaining <= maxFiles else { throw CocoaError(.fileWriteUnknown) }
  }

  private func withDirectoryLock(_ directory: URL, operation: () throws -> Void) throws {
    let descriptor = Darwin.open(directory.appendingPathComponent(".native-runtime.lock").path,
      O_CREAT | O_RDWR, S_IRUSR | S_IWUSR)
    guard descriptor >= 0 else { throw CocoaError(.fileWriteUnknown) }
    defer { Darwin.close(descriptor) }
    let deadline = DispatchTime.now().uptimeNanoseconds + 100_000_000
    while diagnosticFlock(descriptor, LOCK_EX | LOCK_NB) != 0 {
      guard (errno == EWOULDBLOCK || errno == EAGAIN),
            DispatchTime.now().uptimeNanoseconds < deadline else { throw CocoaError(.fileWriteUnknown) }
      usleep(1_000)
    }
    defer { _ = diagnosticFlock(descriptor, LOCK_UN) }
    try operation()
  }

  private static func isJournalFile(_ name: String) -> Bool {
    name.range(of: #"^native-runtime\.(main|keyboard|share)\.[0-9a-f-]{36}\.[0-9]+\.jsonl$"#, options: .regularExpression) != nil
  }

  private static func timestamp(_ date: Date) -> String {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    formatter.timeZone = TimeZone(secondsFromGMT: 0)
    return formatter.string(from: date)
  }

  private static func buildValue(_ key: String) -> String {
    guard let value = Bundle.main.object(forInfoDictionaryKey: key) as? String,
          value.range(of: #"^[0-9A-Za-z.+-]{1,80}$"#, options: .regularExpression) != nil else { return "unknown" }
    return value
  }
}

File: modules/uc-engine/android/src/main/java/expo/modules/ucengine/NativeRuntimeDiagnostics.kt
package expo.modules.ucengine

import android.net.Uri
import android.os.Process
import android.os.SystemClock
import java.io.File
import java.io.RandomAccessFile
import java.nio.channels.FileLock
import java.nio.channels.OverlappingFileLockException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.Semaphore
import java.util.concurrent.TimeUnit
import org.json.JSONObject

internal enum class NativeDiagnosticEvent(val wire: String) {
  CAPTURE_STARTED("capture.started"), APP_FOREGROUND("app.foreground"), APP_BACKGROUND("app.background"),
  ENGINE_START("engine.start"), ENGINE_SUSPEND("engine.suspend"), ENGINE_RESUME("engine.resume"),
  ENGINE_SHUTDOWN("engine.shutdown"), SECURE_STATE_RECOVERY("security.recover"),
  NETWORK_CHANGED("network.path_observed"), NETWORK_OBSERVATION("network.observation"), EXPORT_FLUSH("export.flush")
}

internal enum class NativeDiagnosticTrigger(val wire: String) {
  APP_STARTUP("appStartup"), APP_FOREGROUND("appForeground"), APP_BACKGROUND("appBackground"),
  USER_REQUEST("userRequest"), CONTEXT_DESTROYED("contextDestroyed"), NETWORK_CHANGE("networkChange"),
  EXPORT_REQUESTED("exportRequested"), UNSPECIFIED("unspecified")
}

internal enum class NativeDiagnosticOutcome(val wire: String) {
  OBSERVED("observed"), STARTED("started"), SUCCEEDED("succeeded"), FAILED("failed")
}

internal enum class NativeDiagnosticFailureReason(val wire: String) {
  ENGINE_FAILURE("engineFailure"), PERMISSION_DENIED("permissionDenied"),
  STORAGE_UNAVAILABLE("storageUnavailable"), NATIVE_FAILURE("nativeFailure")
}

internal data class NativeDiagnosticFailure(
  val reason: NativeDiagnosticFailureReason, val code: Long? = null, val retryable: Boolean? = null
)

internal data class NativeDiagnosticNetwork(
  val available: Boolean, val kind: Kind, val expensive: Boolean = false, val constrained: Boolean = false,
  val change: Change = Change.CAPABILITIES
) {
  enum class Change(val wire: String) { INITIAL("initial"), PATH_UPDATE("pathUpdate"), DEFAULT_NETWORK("defaultNetwork"), PROPERTIES("properties"), CAPABILITIES("capabilities") }
  enum class Kind(val wire: String) { WIFI("wifi"), CELLULAR("cellular"), WIRED("wired"), OTHER("other"), NONE("none") }
}

internal data class NativeDiagnosticOperation(
  val id: String, val event: NativeDiagnosticEvent, val trigger: NativeDiagnosticTrigger,
  val startedAt: Long, val generation: Long
)

/** Records only a closed event vocabulary: no caller-provided messages or attribute maps. */
internal class NativeRuntimeDiagnostics(
  private val directory: File?, private val appVersion: String = "unknown", private val appBuild: String = "unknown",
  private val maxFileBytes: Int = 256 * 1024, private val maxFiles: Int = 24
) : AutoCloseable {
  private val sessionId = UUID.randomUUID().toString()
  private val startedAt = Date()
  private val lock = Any()
  private val capacity = Semaphore(128)
  private val queue = Executors.newSingleThreadExecutor { runnable ->
    Thread(runnable, "uc-native-diagnostics").apply { isDaemon = true }
  }
  private var generation = 0L
  private var droppedRecords = 0L
  private var writeFailures = 0L
  private var prunedFiles = 0L
  private var writerReady = false
  private var fileIndex = 0
  private var fileBytes = 0

  init { record(NativeDiagnosticEvent.CAPTURE_STARTED, NativeDiagnosticTrigger.UNSPECIFIED) }

  fun begin(event: NativeDiagnosticEvent, trigger: NativeDiagnosticTrigger): NativeDiagnosticOperation {
    val current = synchronized(lock) {
      if (event == NativeDiagnosticEvent.ENGINE_START) generation++
      generation
    }
    return NativeDiagnosticOperation(UUID.randomUUID().toString(), event, trigger, SystemClock.elapsedRealtime(), current)
      .also { enqueue(event, trigger, NativeDiagnosticOutcome.STARTED, it) }
  }

  fun finish(operation: NativeDiagnosticOperation, outcome: NativeDiagnosticOutcome, failure: NativeDiagnosticFailure? = null) {
    enqueue(operation.event, operation.trigger, outcome, operation, failure)
  }

  fun record(event: NativeDiagnosticEvent, trigger: NativeDiagnosticTrigger,
             outcome: NativeDiagnosticOutcome = NativeDiagnosticOutcome.OBSERVED,
             failure: NativeDiagnosticFailure? = null, network: NativeDiagnosticNetwork? = null) {
    enqueue(event, trigger, outcome, failure = failure, network = network)
  }

  fun flush(deadlineMs: Long = 1000): Boolean {
    val completed = CountDownLatch(1)
    return try {
      queue.execute { completed.countDown() }
      completed.await(deadlineMs.coerceAtLeast(0), TimeUnit.MILLISECONDS)
    } catch (_: RejectedExecutionException) { false }
      catch (_: InterruptedException) { Thread.currentThread().interrupt(); false }
  }

  override fun close() { flush(); queue.shutdown() }

  fun exportSnapshot(): Map<String, Any?> {
    val completed = flush()
    val discovery = discovery()
    return mapOf(
      "flushStatus" to if (completed) "completed" else "incomplete",
      "writer" to snapshot(),
      "fileUris" to discovery.files.map { Uri.fromFile(it).toString() },
      "discoveryStatus" to discovery.status, "skippedFileCount" to discovery.skippedFiles
    )
  }

  fun snapshot(): Map<String, Any?> = synchronized(lock) {
    mapOf(
      "writerStatus" to if (writerReady) "ready" else "unavailable", "droppedRecords" to droppedRecords,
      "writeFailures" to writeFailures, "prunedFiles" to prunedFiles, "sessionId" to sessionId,
      "role" to "main", "startedAt" to timestamp(startedAt), "capturedAt" to timestamp(Date()),
      "policy" to "native-runtime-boundaries-v1", "effectiveLevel" to "info", "filteredRecordCount" to null,
      "retentionDays" to 3, "maxFileBytes" to maxFileBytes, "maxFiles" to maxFiles
    )
  }

  data class Discovery(val files: List<File>, val status: String, val skippedFiles: Int)

  fun files(): List<File> = discovery().files

  fun discovery(): Discovery {
    val candidates = directory?.listFiles() ?: return Discovery(emptyList(), "unavailable", 0)
    var skipped = 0
    val files = candidates.filter {
      if (!it.name.matches(Regex("native-runtime\\.(main|keyboard|share)\\.[0-9a-f-]{36}\\.[0-9]+\\.jsonl"))) return@filter false
      try {
        val regular = it.isFile && it.canonicalFile == File(it.parentFile.canonicalFile, it.name)
        if (!regular) skipped++
        regular
      } catch (_: Exception) { skipped++; false }
    }.sortedBy { it.name }
    return Discovery(files, if (skipped == 0) "completed" else "partial", skipped)
  }

  private fun enqueue(event: NativeDiagnosticEvent, trigger: NativeDiagnosticTrigger, outcome: NativeDiagnosticOutcome,
                      operation: NativeDiagnosticOperation? = null, failure: NativeDiagnosticFailure? = null,
                      network: NativeDiagnosticNetwork? = null) {
    if (!capacity.tryAcquire()) { synchronized(lock) { droppedRecords++ }; return }
    val capturedAt = Date()
    val uptime = SystemClock.elapsedRealtime()
    val currentGeneration = operation?.generation ?: synchronized(lock) { generation }
    try {
      queue.execute {
        try {
          val root = directory ?: error("directory unavailable")
          check(root.isDirectory || root.mkdirs() || root.isDirectory)
          val health = snapshot()
          val record = JSONObject().apply {
            put("schemaVersion", 1); put("policy", "native-runtime-boundaries-v1")
            put("timestamp", timestamp(capturedAt)); put("monotonicMs", uptime)
            put("role", "main"); put("sessionId", sessionId); put("startAttempt", currentGeneration)
            put("processId", Process.myPid()); put("event", event.wire); put("trigger", trigger.wire)
            put("outcome", outcome.wire); put("level", if (outcome == NativeDiagnosticOutcome.FAILED) "ERROR" else "INFO"); put("appVersion", safeBuildValue(appVersion)); put("appBuild", safeBuildValue(appBuild))
            put("droppedRecords", health["droppedRecords"]); put("writeFailures", health["writeFailures"])
            put("prunedFiles", health["prunedFiles"])
            operation?.let {
              put("operationId", it.id)
              if (outcome != NativeDiagnosticOutcome.STARTED) put("durationMs", (uptime - it.startedAt).coerceAtLeast(0))
            }
            failure?.let {
              put("failure", JSONObject().put("reason", it.reason.wire).apply {
                it.code?.let { code -> put("code", code) }
                it.retryable?.let { retryable -> put("retryable", retryable) }
              })
            }
            network?.let {
              put("network", JSONObject().put("available", it.available).put("kind", it.kind.wire)
                .put("expensive", it.expensive).put("constrained", it.constrained).put("change", it.change.wire))
            }
          }
          val bytes = (record.toString() + "\n").toByteArray(Charsets.UTF_8)
          withDirectoryLock(root) {
            if (fileBytes > 0 && fileBytes + bytes.size > maxFileBytes) { fileIndex++; fileBytes = 0 }
            val file = File(root, "native-runtime.main.$sessionId.$fileIndex.jsonl")
            prune(file)
            file.appendBytes(bytes)
            fileBytes += bytes.size
            synchronized(lock) { writerReady = true }
          }
        } catch (_: Exception) {
          synchronized(lock) { droppedRecords++; writeFailures++; writerReady = false }
        } finally { capacity.release() }
      }
    } catch (_: RejectedExecutionException) {
      capacity.release()
      synchronized(lock) { droppedRecords++ }
    }
  }

  private fun prune(current: File) {
    val cutoff = System.currentTimeMillis() - 3 * 24 * 60 * 60 * 1000L
    val discovery = discovery()
    check(discovery.status != "unavailable")
    val candidates = discovery.files.sortedBy { it.lastModified() }
    var remaining = candidates.size + discovery.skippedFiles + if (current in candidates) 0 else 1
    candidates.filter { it != current }.forEach {
      if (it.lastModified() < cutoff || remaining > maxFiles) {
        check(it.delete())
        remaining--
        synchronized(lock) { prunedFiles++ }
      }
    }
    check(remaining <= maxFiles)
  }

  private fun withDirectoryLock(root: File, operation: () -> Unit) {
    RandomAccessFile(File(root, ".native-runtime.lock"), "rw").use { file ->
      var held: FileLock? = null
      val deadline = SystemClock.elapsedRealtime() + 100
      while (held == null) {
        held = try { file.channel.tryLock() } catch (_: OverlappingFileLockException) { null }
        if (held == null) {
          check(SystemClock.elapsedRealtime() < deadline)
          Thread.sleep(1)
        }
      }
      try { operation() } finally { held.release() }
    }
  }

  private fun timestamp(date: Date): String = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.ROOT)
    .apply { timeZone = TimeZone.getTimeZone("UTC") }.format(date)

  private fun safeBuildValue(value: String): String =
    value.takeIf { it.matches(Regex("[0-9A-Za-z.+-]{1,80}")) } ?: "unknown"
}
